File: 1760-minimum-limit-of-balls-in-a-bag/1760-minimum-limit-of-balls-in-a-bag.cpp
class Solution {
public:

bool isPossible(vector<int>& nums, int maxOperations, int penalty) {
    int requiredOps = 0;
    for (int n : nums) {
        int x = n / penalty;
        if (n % penalty == 0) x--;
        requiredOps += x;
    }
    return requiredOps <= maxOperations;
}

int minimumSize(vector<int>& nums, int maxOperations) {
    int start = 1, end = *max_element(nums.begin(), nums.end());
    int minPenalty = end;
    while (start <= end) {
        int penalty = (start+end)/2;
        if (isPossible(nums, maxOperations, penalty))
            minPenalty = penalty, end = penalty-1;
        else
            start = penalty+1;
    }

    return minPenalty;
}

};

File: 1760-minimum-limit-of-balls-in-a-bag/1760-minimum-limit-of-balls-in-a-bag.js
var valid = function (nums, penalty, maxOperations) {
  let requiredOps = 0;
  for (let n of nums) {
    let x = Math.floor(n / penalty);
    if (n % penalty == 0) x--;
    requiredOps += x;
  }
  return requiredOps <= maxOperations;
};
var minimumSize = function (nums, maxOperations) {
  let st = 1,
    end = Math.max(...nums);
  let min_pnatly = Number.MAX_VALUE;
  while (st <= end) {
    let penatly = Math.floor((st + end) / 2);
    if (valid(nums, penatly, maxOperations)) {
      min_pnatly = Math.min(min_pnatly, penatly);
      end = penatly - 1; // because i want to minimize the number
    } else {
      st = penatly + 1;
    }
  }
  return min_pnatly;
};